From 943e815797f886580354f050e7316306c032335c Mon Sep 17 00:00:00 2001
From: Maxim Uvarov <maxim.uvarov@linaro.org>
Date: Wed, 9 Apr 2014 20:38:13 +0400
Subject: [ODP/PATCH] implement odp_timer_disarm_all()

Implement function to disarm all timers. Needed in case of
normal exit from application.
Signed-off-by: Maxim Uvarov <maxim.uvarov@linaro.org>
---
 platform/linux-generic/include/odp_internal.h |  1 +
 platform/linux-generic/source/odp_timer.c     | 24 ++++++++++++++++++++++++
 2 files changed, 25 insertions(+)

diff --git a/platform/linux-generic/include/odp_internal.h b/platform/linux-generic/include/odp_internal.h
index fb3be79..9b0769e 100644
--- a/platform/linux-generic/include/odp_internal.h
+++ b/platform/linux-generic/include/odp_internal.h
@@ -38,6 +38,7 @@ int odp_schedule_init_global(void);
 int odp_schedule_init_local(void);
 
 int odp_timer_init_global(void);
+int odp_timer_disarm_all(void);
 
 #ifdef __cplusplus
 }
diff --git a/platform/linux-generic/source/odp_timer.c b/platform/linux-generic/source/odp_timer.c
index 6fb5025..98ffde3 100644
--- a/platform/linux-generic/source/odp_timer.c
+++ b/platform/linux-generic/source/odp_timer.c
@@ -217,6 +217,30 @@ int odp_timer_init_global(void)
 	return 0;
 }
 
+int odp_timer_disarm_all(void)
+{
+	int timers;
+	struct itimerspec ispec;
+
+	timers = odp_atomic_load_int(&odp_timer.num_timers);
+
+	ispec.it_interval.tv_sec  = 0;
+	ispec.it_interval.tv_nsec = 0;
+	ispec.it_value.tv_sec     = 0;
+	ispec.it_value.tv_nsec    = 0;
+
+	for (; timers >= 0; timers--) {
+		if (timer_settime(odp_timer.timer[timers].timerid,
+				  0, &ispec, NULL)) {
+			ODP_DBG("Timer reset failed\n");
+			return -1;
+		}
+		odp_atomic_fetch_sub_int(&odp_timer.num_timers, 1);
+	}
+
+	return 0;
+}
+
 odp_timer_t odp_timer_create(const char *name, odp_buffer_pool_t pool,
 			     uint64_t resolution, uint64_t min_tmo,
 			     uint64_t max_tmo)
-- 
1.8.5.1.163.gd7aced9

